fog/overseer/server/src/worker.rs

```diff
@@ -0,0 +1,536 @@
+
+impl OverseerWorker {
+    /// Retry failed GRPC requests every 10 seconds.
+    const GRPC_RETRY_SECONDS: Duration = Duration::from_millis(10000);
+
+    pub fn new<DB: RecoveryDb + Clone + Send + Sync + 'static>(
+        ingest_cluster_uris: Vec<FogIngestUri>,
+        recovery_db: DB,
+        logger: Logger,
+        is_enabled: Arc<AtomicBool>,
+    ) -> Self
+    where
+        OverseerError: From<DB::Error>,
+    {
+        let thread_is_enabled = is_enabled;
+        let stop_requested = Arc::new(AtomicBool::new(false));
+        let thread_stop_requested = stop_requested.clone();
+        let grpcio_env = Arc::new(grpcio::EnvBuilder::new().build());
+        let ingest_clients: Vec<FogIngestGrpcClient> = ingest_cluster_uris
+            .iter()
+            .map(|fog_ingest_uri| {
+                FogIngestGrpcClient::new(
+                    fog_ingest_uri.clone(),
+                    Self::GRPC_RETRY_SECONDS,
+                    grpcio_env.clone(),
+                    logger.clone(),
+                )
+            })
+            .collect();
+        let join_handle = Some(
+            ThreadBuilder::new()
+                .name("OverseerWorker".to_string())
+                .spawn(move || {
+                    OverseerWorkerThread::start(
+                        ingest_clients,
+                        recovery_db,
+                        thread_is_enabled,
+                        thread_stop_requested,
+                        logger,
+                    )
+                })
+                .expect("Could not spawn OverseerWorkerThread"),
+        );
+
+        Self {
+            join_handle,
+            stop_requested,
+        }
+    }
+
+    /// Stop and join the db poll thread
+    pub fn stop(&mut self) -> Result<(), ()> {
+        if let Some(join_handle) = self.join_handle.take() {
+            self.stop_requested.store(true, Ordering::SeqCst);
+            join_handle.join().map_err(|_| ())?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Drop for OverseerWorker {
+    fn drop(&mut self) {
+        let _ = self.stop();
+    }
+}
+
+/// The thread that performs the Fog Overseer logic.
+struct OverseerWorkerThread<DB: RecoveryDb> {
+    /// The list of FogIngestClients that Overseer uses to communicate with
+    /// each node in the Fog Ingest cluster that it's monitoring.
+    ingest_clients: Vec<FogIngestGrpcClient>,
+
+    /// The database that contains, among other things, info on the Fog Ingest
+    /// cluster's ingress keys.
+    recovery_db: DB,
+
+    /// If this is true, the worker will not perform it's monitoring logic.
+    is_enabled: Arc<AtomicBool>,
+
+    /// If this is true, the thread will stop.
+    stop_requested: Arc<AtomicBool>,
+
+    logger: Logger,
+}
+
+/// This associates an IngestSummary with an IngestClient. This makes it easy
+/// to query a given node based on its IngestSummary.
+struct IngestSummaryNodeMapping {
+    node_index: usize,
+
+    ingest_summary: IngestSummary,
+}
```

Comment:
yeah that would be more compact, and earlier I wasn't sure if I should be adding these derives to IngestSummary.

That said, I weakly prefer keeping it as is because it allows us to name what the `usize` parameter is. Without this named `node_index` field, I could foresee readers having a hard time understanding what the usize means. That said, I'd be happy to change it. LMK.

Reply:
I don't have a strong preference, how often are queries (map) vs. iterations (vec) is really the determining factor...

Reply:
Good point. We're doing more iterations than queries, so will leave it as such.